### PR TITLE
Added more tests for `Drop`

### PR DESCRIPTION
**Pull request by YoshikiTakashima**
### Description of changes: 

Kani's behavior does not change. 4 tests are added to test Kani's behavior on the `Drop` trait. These test enums, channels, `Rc<Refcell<T>>` and ownership-moving closures.

### Resolved issues: 
Issue #1284

### Call-outs:
1. The channel test case (`tests/kani/Drop/fixme_drop_after_moving_across_channel.rs`) fails when `kani::unwind(1)` is set, and does not finish in under 10m when `kani::unwind(2)` is set. This will get a separate issue once merged.
2. I've placed the tests in places that I think will fit, but please object if you think the tests should be somewhere else.

### Testing
* How is this change tested? All new test cases run with `scripts/kani-regression.sh`, also manually check that assertions are valid.

* Is this a refactor change? No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
  - Similar level as other test cases.
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by danielsn — COMMENTED** [review]
LGTM, modulo comments. 

Does this resolve all the open `drop` testing issues?

@adpaco-aws thoughts on the file name/locations for these tests?

File: tests/kani/Closure/drop_after_move_closure_call.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// This test checks that structs moved into Box<Fn> is dropped when
```
Comment (adpaco-aws) [inline tests/kani/Closure/drop_after_move_closure_call.rs:4]:
```suggestion
// This test checks that structs moved into Box<Fn> are dropped when
```

File: tests/kani/Enum/drop_enum_only_one_called.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+}
+
+enum EnumWithTwoIncrements {
+    Add1(IncrementCELLWhenDropped),
+    Add2(IncrementCELLWhenDropped),
+}
+
+fn get_random_enum_variant(random: u32) -> EnumWithTwoIncrements {
+    if random % 2 == 0 {
```
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply tests/kani/Enum/drop_enum_only_one_called.rs:26]: Or take no arguments and use `kani::any()` as the condition.

[the remark this answers is not in the document]
Comment (YoshikiTakashima) [reply tests/kani/Enum/drop_enum_only_one_called.rs:26]: No. I'll patch.

File: tests/kani/Drop/fixme_drop_after_moving_across_channel.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// This test checks whether dropping objects passed through
+// std::sync::mpsc::channel is handled.
+//
+// This test case fails to resolve in a reasonable amount of
```
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply tests/kani/Drop/fixme_drop_after_moving_across_channel.rs:7]: +1, we normally create the issue first then paste the link below.

**Review by adpaco-aws — COMMENTED** [review]
I'd rather have all drop-related tests in the `Drop` folder so we have them all in one place. As long as their names refer to the types for which it's being tested (which they do), they're fine.

File: tests/kani/Drop/fixme_drop_after_moving_across_channel.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (YoshikiTakashima) [reply tests/kani/Drop/fixme_drop_after_moving_across_channel.rs:7]: Opened #1286

File: tests/kani/Drop/fixme_drop_after_moving_across_channel.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// This test checks whether dropping objects passed through
```
[the remark this answers is not in the document]
Comment (YoshikiTakashima) [reply tests/kani/Drop/fixme_drop_after_moving_across_channel.rs:4]: `//!` is for to document the "enclosing item," I believe that is true. I would patch for this one.

I don't think that was enforced for a large portion of the test cases under `tests/kani`. Recommend separate cleanup.

File: tests/kani/Drop/drop_after_mutating_refcell.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        unsafe {
+            CELL = self.get_inner_value();
+        }
+    }
+}
+
+#[kani::proof]
+fn main() {
+    {
+        let set_to_one = DropSetCELLToInner { set_cell_to: 1 };
```
[the remark this answers is not in the document]
Comment (YoshikiTakashima) [reply tests/kani/Drop/drop_after_mutating_refcell.rs:47]: Resolved with 5f5203b52fd83eb9e10babe619f55164d75ff832

**Comment by YoshikiTakashima** [comment]
@adpaco-aws @danielsn:

- Drop related tests are placed under `tests/kani/Drop`.
- I'm not too familiar with the custom drop behaviors that are built in, but the existing set (including previous test cases) should cover the areas identified at the start of this week.

**Review by adpaco-aws — APPROVED** [review]
Thanks!

**Review by danielsn — APPROVED** [review]

**Pull request merged**